src/compiler/checker.ts

```diff
@@ -3690,6 +3697,48 @@ module ts {
             }
             return links.resolvedType;
         }
+        
+        function getTypeGuardType(node: TypeGuardTypeNode): Type {
+            let type = <TypeGuardType>createType(TypeFlags.TypeGuard);
+            let hasParameters = false;
+            type.parameterIndex = -1;
+            if (node.target) {
+                let functionDeclaration = <SignatureDeclaration>node.target.parent.parent;
+                if (functionDeclaration.parameters) {
+                    for (let i = 0; i < functionDeclaration.parameters.length; i++) {
+                        let param = functionDeclaration.parameters[i];
+                        if (param.name.kind === SyntaxKind.Identifier && (<Identifier>param.name).text === node.target.text) {
+                            type.parameterIndex = i;
+                            break;
+                        }
+                    }
+                    hasParameters = true;
+                }
+                type.parameterName = node.target.text;
+            }
+            if (hasParameters && type.parameterIndex === -1) {
+                error(node.target, Diagnostics.Type_guard_target_must_have_a_matching_parameter);
+            }
+            if (node.type) {
+                type.type = getTypeFromTypeNode(node.type);
+                if (type.type.flags & TypeFlags.TypeGuard) {

```

Comment:
This only handles a type guard directly inside a type guard. If it is nested more deeply, I guess it will not be caught.